Circle variants for drawing modes

fsobolev opened the pull request:
PR to track progress of implementing Circle variants for drawing modes

- [x] Wave
- [x] <del>Line</del> (removed in #38)
- [ ] Levels
- [ ] Particles
- [ ] Spine
- [x] Bars

Related issue: #7

fsobolev committed:
Add circle option

fsobolev committed:
Implement Line mode Circle variant

fsobolev committed:
Implement radius setting

fsobolev committed:
Radial gradient

fsobolev committed:
Wave mode Circle variant

ghost commented:
Nice

ghost commented:
Spine could actually look *really* cool in circle mode.

ghost commented:
I think Wave should be filled completely in circle instead of having a blank circle on the inside. We could think about introducing a setting for that, but I'm not sure if that's worth it.

ghost commented:
Looks amazing otherwise!

fsobolev commented:
> I think Wave should be filled completely in circle instead of having a blank circle on the inside. We could think about introducing a setting for that, but I'm not sure if that's worth it.

The setting would be a good compromise. Probably called "wave-circle-fill" in the schema, and in the UI it would be a switch with description "Fill completely", it can be added as the suffix of Wave row and only shown when Circle is on.

fsobolev committed:
Wave Circle Fill Option

fsobolev commented:
![](https://i.imgur.com/UyqwTiz.png)

Wave Circle Fill option added. It's not shown when Box variant is selected.

fsobolev committed:
Fix radial gradient

fsobolev committed:
Bars mode Circle variant

fsobolev committed:
Bars fill setting

fsobolev committed:
Wave Circle - Fix min radius when not filled

fsobolev commented:
It's probably can be expected from Circle Bars mode to have bars with constant width... But IMO this looks way cooler

| ![](https://i.imgur.com/Aqp1GGq.png) | ![](https://i.imgur.com/VsrOB4t.png) |
|---|---|
| ![](https://i.imgur.com/vm9uV2X.png) | ![](https://i.imgur.com/P0LSvYL.png) |

ghost commented:
I can look into the code and see if I can work on some of the other modes. Problem is, I don't know anything about those Sinus and Cosinus functions.

fsobolev commented:
> Problem is, I don't know anything about those Sinus and Cosinus functions.

They are used to get coordinates of a point on a circle. The formula is:
```
x = cos(angle) * radius
y = sin(angle) * radius
```

Let's see how it works with Bars. The x coordinate of the first point of a bar is calculated like this:
```
width / 2 + math.cos(2 * math.pi / ls * (i + offset / 100) - 0.5 * math.pi) * min_radius,
```
First we need to find an angle in [radians](https://en.wikipedia.org/wiki/Radian). `2 * math.pi` is the full circle (= 360 deg). It's divided by number of bars, and then multiplied by the index of current bar + offset. Imagine like you're slicing a pizza, let's say you want to share it across 12 people, but you can only cut from center to sides, so you need to do 12 cuts, each one rotated relatively to previous one by fixed angle, that's basically how it works 😄 
`- 0.5 * math.pi` (- 90 deg) is used to rotate everything so the angle 0 will be on top of the circle instead of the right.
So now we have an angle, it's time to get it's cosinus and then multiply it by radius.
But since we want to draw our circle from the center of the area instead of the beginning of coordinates, `width / 2` is added.
y coordinate is calculated the same, but using height and sinus instead of width and cosinus respectively.

Hope it helps you to understand how this works :)

ghost commented:
Thanks! I'll see what I can do from there.

fsobolev committed:
Rework radius setting

fsobolev committed:
Impove draw_element, particles and spine

draw_element now acceps center coordinates instead of top left corner, making it easier to use with circle variants.
Also gradients are now better applied in Spine mode

fsobolev committed:
Fix particles mode

It was broken after previous commit

fsobolev committed:
Remove bars-fill setting and use fill instead

fsobolev committed:
Make Bars Circle mode more or less working

fsobolev commented:
All changes made in #38 were merged here and I made Bars Circle mode working with proper use of offset and thickness.
There are 2 problems with this mode tho:
1. To avoid really broken look, thickness was limited, but I didn't find good maximum value yet, currently it's limited to `circle_length / number_of_bars * 0.25` which makes it maximum 12 when maximized (on 1360x768 display), so it's not very good.
2. CPU usage is insane due to complex calculation of angles depending on thickness. It uses 7-10% more CPU for me compared to the same settings with Boxed variant. I hope to at least partially fix it by moving some calculations out of the loop to calculate most numbers once per frame instead of for each bar, I will do it tomorrow.

fsobolev committed:
Improve box/circle preferences

Commit:
add line circle mode to wave circle mode

Commit:
improve & rename Wave mode fill circle setting

Commit:
Add shortcut for inner circle & more renaming
